### PR TITLE
Live Tests: Stabilization and Resilience (Phase 1)

### PR DESCRIPTION
**Pull request by jsquire**
## Summary

The Service Bus client library tests which were designated as Live are not fully isolated nor resilient, resulting in some unpredictable failures and, over time, also accumulating consistent failures over time as part of the nightly runs. 

In their current form, the tests make use of a set of shared resources under a given Service Bus namespace, with multiple tests making use of the same queue/topic/subscription.  Additionally, the tests are not always robust and may leave orphaned messages or resources when an unexpected exception or test failure occurs.  

As a result, the result of one test can influence the outcome of tests that come after, causing failures to cascade.  Further, there are some cases where the tests make assumptions about the state of the global environment, which may or may not hold true depending on the order in which tests are run and the results of test previously executed.

## Goals
- Increase reliability of the tests during nightly runs, eliminating known failures and adding stability to tests whose results are non-deterministic.

- Isolate tests from one another, ensuring that they are fully self-contained with no dependencies on each other or a specific state of the global environment.  

- Increase test resiliency; tests should make every reasonable attempt to clean up after themselves, removing any changes to the environment that they added.  Where it is reasonable to do so, test setup/teardown should anticipate potential transient issues and retry operations.  

- Remove the assumption of exclusive shared state access.  As part of the efforts to isolate tests and allow them to run independently, the tests should not use an implementation which is based on the assumption that only one test is ever executing and that it owns both the local and Azure environments exclusively.

- Add polish to test infrastructure, ensuring that tests have a common base of utilities to work with, and which can clearly and easily be identified as such. 

- Attempt to minimize changes made, keeping the original approach, test cases, code styles

## Non-Goals
- Refactoring of the general approach for testing, the test cases, code style, and other artifacts of the codebase; changes should be kept as tightly scoped and as minimal as possible.

- Adding of additional test cases or new testing scenarios.

- Refactoring of the reliance on timing for test success; _this will be a goal in the second phase of these efforts._

- Allowing test retries in order to better deal with the non-determinism in message timings and/or transient network issues; _this will be a goal in the second phase of these efforts._

- Enabling test parallelism; _this will be a stretch goal in the second phase of these efforts._

### General

- Renamed environment variable to something better aligned with the SDKs for other Azure services.

- Updated ARM template to remove creation of the global shared queues, topics, and subscriptions.

- Introduced new infrastructure to allow for topics/queues scoped to a given test.  This was intended to allow tests to be sandboxed so that they were unable to impact one another, preventing cascading failures or state corruption.

- Reorganized some test support files, to separate them from the tests themselves and decrease clutter.

- Refactored how environment variables are read, removing reliance on a static constructor and allowing test classes to use the `TestUtility` outside of a live test context.

### TestUtility

- Move to infrastructure for better organization  
  
- Remove static constructor in favor of lazy read/validation of environment variables on first use.
  
- Allows use of utility functionality for non-live tests when environment variable is not present.

- Making the logging methods silently swallow errors, as logging failures should not be fatal.  

### Test Constants

- Removed unused members  
  
- Removed named queues/topics, as there are no more global items  

- Organized by category and added clarifying comments

### Live Tests

- Heavy refactoring to remove implicit parameter passing via abstract class members, as they hurt readability and isolation and blocked parallel execution.

- Refactored management client tests in order to introduce infrastructure to cleanup resources potentially orphaned by test failures or unexpected exceptions.

- Refactoring of tests to remove use of shared global resources within the Service Bus namespace and to make use of the new infrastructure to create/delete those needed for an individual test.  

- Adding a helper for safe closing of entities in TransactionalTests to minimize  finally bloat.  

- Refactoring of the timeout pattern to ensure that the timeout task is cancelled so that it isn't leaked in WebSocketsEnd2EntTests.  

## Last Upstream Rebase

Wednesday, March 27, 2019 06:16pm (EDT)

## Related and Follow-Up Issues

- [[Service Bus Client] Stabilize Live Tests](https://github.com/Azure/azure-sdk-for-net/issues/5446) (#5446)

### CODE REVIEW

**Comment by jsquire** [comment]
@mikeharder: If you don't mind, I could use a bit of advice.   The build doesn't seem to be picking up the environment variable change from `client.yml`.   Builds are failing with:  `Error Message:
 System.InvalidOperationException : 'AZ_SERVICE_BUS_CONNECTION' environment variable was not found!`

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/API/ApprovalTestConfig.cs [hunk 1]
```diff
@@ -1,7 +0,0 @@
-﻿using ApprovalTests.Reporters;
```
Comment (jsquire) [inline src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/API/ApprovalTestConfig.cs]: This is a weird diff.   The file was moved to the folder `AssemblyConfiguration` without changes.  Unsure why this appears as it does.

**Review by pakrym — COMMENTED** [review]
I like the direction :+1:.

I'm not a fan of blocking Dispose in a lot of these test types and mixed lifetimes where clients are created in lambda stored into fields and Disposed by the test destructor but it might be something to fix in the future.

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/QueueClientDiagnosticsTests.cs [hunk 2]
```diff
@@ -12,443 +12,463 @@ namespace Microsoft.Azure.ServiceBus.UnitTests.Diagnostics
...
-                    processActivity = Activity.Current;
-                    processingDone.Set();
-                    return Task.CompletedTask;
-                },
-                exArgs => Task.CompletedTask);
-
-            processingDone.WaitOne(TimeSpan.FromSeconds(maxWaitSec));
-            Assert.True(this.events.IsEmpty);
-            Assert.Null(processActivity);
+            await ServiceBusScope.UsingQueueAsync(partitioned: false, sessionEnabled: false, async queueName =>
```
[the remark this answers is not in the document]
Comment (pakrym) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/QueueClientDiagnosticsTests.cs]: I know it doesn't work in all the cases but a would simplify a lot of them.

[the remark this answers is not in the document]
Comment (jsquire) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/QueueClientDiagnosticsTests.cs]: I'm not a fan either, honestly.   That said, I looked at changing it and decided against doing so to avoid refactoring the `Dispose` approach for closing and cleaning things up.    Given the wide scope of changes, I'm inclined to just leave this as-is.

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/QueueClientDiagnosticsTests.cs [hunk 3]
```diff
@@ -12,443 +12,463 @@ namespace Microsoft.Azure.ServiceBus.UnitTests.Diagnostics
...
+            {                
+                this.queueClient = new QueueClient(TestUtility.NamespaceConnectionString, queueName, ReceiveMode.ReceiveAndDelete);
+
+                Activity processActivity = null;
+
+                ManualResetEvent processingDone = new ManualResetEvent(false);
```
[the remark this answers is not in the document]
Comment (jsquire) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/QueueClientDiagnosticsTests.cs]: /agree.   I've got this flagged for follow-up in phase 2 of the changes.

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Management/ManagementClientTests.cs [hunk 4]
```diff
@@ -27,61 +28,69 @@ public async Task BasicQueueCrudTest()
...
+                await Assert.ThrowsAsync<MessagingEntityNotFoundException>(
+                        async () =>
+                        {
+                            await client.GetQueueAsync(qd.Path);
+                        });
+
+                exists = await client.QueueExistsAsync(queueName);
+                Assert.False(exists);
+            }
+            catch
```
[the remark this answers is not in the document]
Comment (jsquire) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Management/ManagementClientTests.cs]: These tests are not running in the scope of a given Service Bus resource, they're testing the operations for creating and managing them.   The lambda scope elsewhere is responsible for setting up the resource that the test expects so that it runs in the scope.... I gave a look do doing something similar here, but it seemed over complicated, so I backed it out and went simple try/catch.

With respect to `catch` over `finally`, the tests do their own cleanup and then assert against that in the normal flow, so unless there's a failure or unexpected exception, the additional code that I added shouldn't have to run.    But, these have left orphaned resources in the Service Bus namespace when they did have a failure/exception, so I added an additional layer of cleanup for that corner case.

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Management/ManagementClientTests.cs [hunk 5]
```diff
@@ -663,5 +774,29 @@ public void Dispose()
...
+
+        private async Task SafeDeleteQueue(string name, [CallerMemberName] string caller = null)
+        {
+            try
+            {
+                await client?.DeleteQueueAsync(name);
+            }
+            catch (Exception ex)
+            {
+                try { TestUtility.Log($"{ caller } could not delete the queue [{ name }].  Error: [{ ex.Message }]"); } catch {}
```
[the remark this answers is not in the document]
Comment (jsquire) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Management/ManagementClientTests.cs]: Yeah.... that would be the smarter thing to do, wouldn't it.    Will change.

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/AmqpConverterTests.cs [hunk 6]
```diff
@@ -139,7 +133,6 @@ void SB_message_with_no_TTL_results_in_empty_Ampq_TTL()
         }
 
         [Fact]
-        [LiveTest]
```
Comment (AlexGhiondea) [inline src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/AmqpConverterTests.cs]: I assume we removed this because it doesn't hit the live server, right?

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/WebSocketsEnd2EndTests.cs [hunk 7]
```diff
@@ -15,41 +15,45 @@ public sealed class WebSocketsEnd2EndTests
...
+                            return Task.CompletedTask;
+                        });
+                    await queueClient.SendAsync(new Message(contentAsBytes));
+
+                    var timeoutTask = Task.Delay(Timeout);
+                    var receiveTask = taskCompletionSource.Task;
 
-                var timeoutTask = Task.Delay(Timeout);
-                var receiveTask = taskCompletionSource.Task;
+                    if (await Task.WhenAny(timeoutTask, receiveTask) == timeoutTask)
```
[the remark this answers is not in the document]
Comment (jsquire) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/WebSocketsEnd2EndTests.cs]: Good catch!   Taking note of this for the second phase.   Will address it there.

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/AmqpConverterTests.cs (hunk 6 above)
Comment (jsquire) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/AmqpConverterTests.cs]: Correct.   There were a bunch of things that were bound to `TestUtility` which had dependencies on the environment variable due to static initialization.   I tweaked the approach for initialization and was able to reclassify ~15 tests that weren't actually `Live`.

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Infrastructure/ServiceBusScope.cs [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+        /// <param name="sessionEnabled">If <c>true</c>, a session will be enabled on the queue.</param>
+        /// <param name="configureQueue">If provided, an action that can override the default properties used for queue creation.</param>
+        /// <param name="caller">The name of the calling method; this is intended to be populated by the runtime.</param>
+        /// 
+        /// <returns>The queue scope that was created.</returns>
+        /// 
+        public static async Task<QueueScope> CreateQueueAsync(bool partitioned, 
+                                                              bool sessionEnabled,                                                   
+                                                              Action<QueueDescription> configureQueue = null, 
+                                                              [CallerMemberName] string caller = "")
```
Comment (AlexGhiondea) [inline src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Infrastructure/ServiceBusScope.cs]: Neat!

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Infrastructure/ServiceBusScope.cs [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+            var topicDescription = BuildTopicDescription(topicName, partitioned);
+            var subscriptionDescription = BuildSubscriptionDescription(subscripionName, topicName, sessionEnabled);
+
+            configureTopic?.Invoke(topicDescription);
+            configureSubscription?.Invoke(subscriptionDescription);
+
+            await CreateRetryPolicy<TopicDescription>().ExecuteAsync( () => ManagementClient.CreateTopicAsync(topicDescription));
+            await CreateRetryPolicy<SubscriptionDescription>().ExecuteAsync( () => ManagementClient.CreateSubscriptionAsync(subscriptionDescription));
+                 
+
```
Comment (AlexGhiondea) [inline src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Infrastructure/ServiceBusScope.cs]: nit: extra line

**Review by AlexGhiondea — APPROVED** [review]

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Infrastructure/ServiceBusScope.cs (hunk 9 above)
Comment (jsquire) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Infrastructure/ServiceBusScope.cs]: Whoops.  Got it.

**Comment by jsquire** [comment]
> I'm not a fan of blocking Dispose in a lot of these test types and mixed lifetimes where clients are created in lambda stored into fields and Disposed by the test destructor but it might be something to fix in the future.

I agree 100%.   I'm not a fan of the approach that was used for managing resources and would have preferred the tests themselves handled cleanup, since they're already `async` and don't need to block.  I'll file an issue for the backlog to track potentially revisiting in the future, in case the legacy client sticks around for a while.

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/TransactionTests.cs [hunk 10]
```diff
@@ -527,13 +566,20 @@ public async Task TransactionalSendViaCommitTest()
...
             finally
             {
-                await intermediateSender.CloseAsync();
-                await intermediateReceiver.CloseAsync();
-                await destination1Sender.CloseAsync();
-                await destination1ViaSender.CloseAsync();
-                await destination2ViaSender.CloseAsync();
-                await destination1Receiver.CloseAsync();
-                await destination2Receiver.CloseAsync();
+                try { await intermediateSender.CloseAsync();    } catch {}
```
[the remark this answers is not in the document]
Comment (jsquire) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/TransactionTests.cs]: Interesting idea.   Tweak incoming.

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/WebSocketsEnd2EndTests.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (jsquire) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/WebSocketsEnd2EndTests.cs]: @pakrym:  Updated.  I think that I've got the pattern correct.  Would appreciate a double-check.

[the remark this answers is not in the document]
Comment (pakrym) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/WebSocketsEnd2EndTests.cs]: Looks good
